### ISSUE TITLE
Generated models don't generate great code for string enums

### ISSUE THREAD

**Issue by kojustin** [go-swagger/go-swagger#861]
## Problem statement

When generating code for fields that are enums, the generated code is poor.

1. The validation logic needs an "init" function to populate a global singleton. The init function does an json.Unmarshal to extract the values it wants to compare against. However, the values are constants that are embedded into the swagger spec. The init function and this global could be replaced by referring to the constants StatusStatusBad, StatusStatusHorrible, etc. which are already generated.

2. It shouldn't need to import "github.com/go-openapi/validate" to validate a string is one of the enum values. Requiring another dependency for this is kind of lame.

3. It would be nice if the enum introduced a new type alias. This helps preserve the intention behind the enum (to separate from the underlying type, to limit the possible set of values).

```
type Status string

const (
	// StatusStatusGood captures enum value "Good"
	StatusStatusGood Status = "Good"
	// StatusStatusBad captures enum value "Bad"
	StatusStatusBad = "Bad"
	// StatusStatusOk captures enum value "Ok"
	StatusStatusOk = "Ok"
	// StatusStatusDecent captures enum value "Decent"
	StatusStatusDecent = "Decent"
	// StatusStatusMiddling captures enum value "Middling"
	StatusStatusMiddling = "Middling"
	// StatusStatusHorrible captures enum value "Horrible"
	StatusStatusHorrible = "Horrible"
)
```

## Swagger specification
```
definitions:
  Status:
    type: object
    description: |
      some object
    properties:
      status:
        description: status
        type: string
        enum:
          - Good
          - Bad
          - Ok
          - Decent
          - Middling
          - Horrible
```

## Steps to reproduce
```
$ /usr/local/Cellar/go-swagger/0.8.0/bin/swagger generate model --spec=swag.yml
$ cat models/status.go 
package models

// This file was generated by the swagger tool.
// Editing this file might prove futile when you re-run the swagger generate command

import (
	"encoding/json"

	strfmt "github.com/go-openapi/strfmt"
	"github.com/go-openapi/swag"

	"github.com/go-openapi/errors"
	"github.com/go-openapi/validate"
)

// Status some object
//
// swagger:model Status
type Status struct {

	// status
	Status string `json:"status,omitempty"`
}

// Validate validates this status
func (m *Status) Validate(formats strfmt.Registry) error {
	var res []error

	if err := m.validateStatus(formats); err != nil {
		// prop
		res = append(res, err)
	}

	if len(res) > 0 {
		return errors.CompositeValidationError(res...)
	}
	return nil
}

var statusTypeStatusPropEnum []interface{}

func init() {
	var res []string
	if err := json.Unmarshal([]byte(`["Good","Bad","Ok","Decent","Middling","Horrible"]`), &res); err != nil {
		panic(err)
	}
	for _, v := range res {
		statusTypeStatusPropEnum = append(statusTypeStatusPropEnum, v)
	}
}

const (
	// StatusStatusGood captures enum value "Good"
	StatusStatusGood string = "Good"
	// StatusStatusBad captures enum value "Bad"
	StatusStatusBad string = "Bad"
	// StatusStatusOk captures enum value "Ok"
	StatusStatusOk string = "Ok"
	// StatusStatusDecent captures enum value "Decent"
	StatusStatusDecent string = "Decent"
	// StatusStatusMiddling captures enum value "Middling"
	StatusStatusMiddling string = "Middling"
	// StatusStatusHorrible captures enum value "Horrible"
	StatusStatusHorrible string = "Horrible"
)

// prop value enum
func (m *Status) validateStatusEnum(path, location string, value string) error {
	if err := validate.Enum(path, location, value, statusTypeStatusPropEnum); err != nil {
		return err
	}
	return nil
}

func (m *Status) validateStatus(formats strfmt.Registry) error {

	if swag.IsZero(m.Status) { // not required
		return nil
	}

	// value enum
	if err := m.validateStatusEnum("status", "body", m.Status); err != nil {
		return err
	}

	return nil
}

```

**Comment by casualjim** [comment]
if you want a type alias you shouldn't use an inline definition of the enum, but move the enum to a definition itself. That will hint the code generator that you want it to be a type alias. (takes care of issue 3 in your list)  I wouldn't object to expanding this to also do that for inline enum definitions.

I think there are many different kinds of enums to be compliant with json schema and string enums are not the only ones that matter.  According to json-schema _any_ type can be used in an enum this includes arrays and objects

Not importing of validate is kind of futile because other validations that are present will require it anyway. Surely your application doesn't have 1 model with 1 property that is an enum so I don't quite understand the objection to it. If it is 1 model with 1 property and that is an enum I don't know what codegen buys you at all.

The reason it uses a singleton is because of complex objects like nested structs can also be enums in which case it's much harder to generate constants for them. init comes at a negligible one time cost so other than cleanliness I don't know what the objection is as this is unexported usage in generated code. So what is the objection besides it's not the code you would have preferred to see the most for this narrower use case.

**Comment by GlenDC** [comment]
+1 to expand the type alias behavior to inline enum definitions. @casualjim I'll open a new issue for this, as it seems to deviate from this context.

**Comment by fredbi** [comment]
Preparing a PR which solves quite a few issues with enum constant generation and gives more control about constant names, for any underlying enum type. It allows for const naming override with a "x-go-const-names" custom extension.

As a by-product, this development solves (1) in favourable cases (e.g. generating cleaner code when enum is made of simple types, trying hard to generate const whenever possible).

I don't think (2) is something we want to add. The validation method is clean and reusable. One might object that it is slower because it uses type reflection. Anyway, I am not addressing this ATM.

I consider too that (3) is addressed by putting the _intent_ in the specification. If anyone is interested, I may add an option to export the generated enum slice (currently used internally for validation).

**Comment by gregmarr** [comment]
Is (3) what's requested in #868 and #1328 ?

**Comment by fredbi** [comment]
Yeah, but Glen has set even higher expectations: to figure out the enum stuff for all things, even from SimpleSchema objects. That sounds nice, but I am not going to implement such a big change.

**Comment by fredbi** [comment]
My current dev is trying to make enum const customizable and performs some sanity check before generating const. I took some time too to generate more idiomatic initializers and enable const generation on _any_ enum type, not only strings.
I think this kind of feature could be reused later on to do something similar on default values.

**Comment by fredbi** [comment]
This is a sample output from my dev. Is that better code? (note the generalization to ANY type).
Obviously this is a favourable case. For more complex structures, a var + init() with marshaller is unavoidable.

From spec:
```yaml
  bossId:
    type: string
    format: uuid
    enum:
      - 'a8098c1a-f86e-11da-bd1a-00112444be1e'
      - 'a8098d1a-f86e-11da-bd1a-00112444be1e'
      - 'a8098e1b-f86e-11db-bd1a-00112444be1e'
    x-go-const-names:
      - BossMasterID
      - HeadMasterID
      - ChiefID
```
```golang
// BossID boss Id
// swagger:model bossId
type BossID strfmt.UUID

const (
	// BossMasterID captures enum value a8098c1a-f86e-11da-bd1a-00112444be1e
	BossMasterID = strfmt.UUID("a8098c1a-f86e-11da-bd1a-00112444be1e")

	// ChiefID captures enum value a8098e1b-f86e-11db-bd1a-00112444be1e
	ChiefID = strfmt.UUID("a8098e1b-f86e-11db-bd1a-00112444be1e")

	// HeadMasterID captures enum value a8098d1a-f86e-11da-bd1a-00112444be1e
	HeadMasterID = strfmt.UUID("a8098d1a-f86e-11da-bd1a-00112444be1e")
)

// for schema
const bossIdEnum = []BossID{BossMasterID, HeadMasterID, ChiefID}

func (m BossID) validateBossIDEnum(path, location string, value BossID) error {

	if err := validate.Enum(path, location, value, bossIdEnum); err != nil {
		return err
	}
	return nil
}
```

**Comment by HarlamovBuldog** [comment]
> This is a sample output from my dev. Is that better code? (note the generalization to ANY type).
> Obviously this is a favourable case. For more complex structures, a var + init() with marshaller is unavoidable.
> 
> From spec:
> 
> ```yaml
>   bossId:
>     type: string
>     format: uuid
>     enum:
>       - 'a8098c1a-f86e-11da-bd1a-00112444be1e'
>       - 'a8098d1a-f86e-11da-bd1a-00112444be1e'
>       - 'a8098e1b-f86e-11db-bd1a-00112444be1e'
>     x-go-const-names:
>       - BossMasterID
>       - HeadMasterID
>       - ChiefID
> ```
> 
> ```go
> // BossID boss Id
> // swagger:model bossId
> type BossID strfmt.UUID
> 
> const (
> 	// BossMasterID captures enum value a8098c1a-f86e-11da-bd1a-00112444be1e
> 	BossMasterID = strfmt.UUID("a8098c1a-f86e-11da-bd1a-00112444be1e")
> 
> 	// ChiefID captures enum value a8098e1b-f86e-11db-bd1a-00112444be1e
> 	ChiefID = strfmt.UUID("a8098e1b-f86e-11db-bd1a-00112444be1e")
> 
> 	// HeadMasterID captures enum value a8098d1a-f86e-11da-bd1a-00112444be1e
> 	HeadMasterID = strfmt.UUID("a8098d1a-f86e-11da-bd1a-00112444be1e")
> )
> 
> // for schema
> var bossIdEnum = []BossID{BossMasterID, HeadMasterID, ChiefID}
> 
> func (m BossID) validateBossIDEnum(path, location string, value BossID) error {
> 
> 	if err := validate.Enum(path, location, value, bossIdEnum); err != nil {
> 		return err
> 	}
> 	return nil
> }
> ```

Is this feature already implemented or how long will it take?
